### PR TITLE
Add approvaltests.cpp 10.12.0

### PR DESCRIPTION
**Pull request by isidore**
Specify library name and version:  **approvaltests.cpp/10.12.0**

new version

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by madebr — APPROVED** [review]

**Comment by claremacrae** [comment]
This build is failing, but not due to anything in the PR… I think it’s likely to be due to a change in the Conan CI config?

boost 1.72.0 is missing for one config: <https://c3i.jfrog.io/c3i/misc/logs/pr/7213/1-configs/macos-m1-clang/approvaltests.cpp/10.12.0//5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9-test.txt> … we didn’t change anything in that area….

I’ve asked about it in the Conan channel on #include <C++> Discord.

**Comment by SSE4** [comment]
> This build is failing, but not due to anything in the PR… I think it’s likely to be due to a change in the Conan CI config?
> 
> boost 1.72.0 is missing for one config: https://c3i.jfrog.io/c3i/misc/logs/pr/7213/1-configs/macos-m1-clang/approvaltests.cpp/10.12.0//5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9-test.txt … we didn’t change anything in that area….
> 
> I’ve asked about it in the Conan channel on #include <C++> Discord.

yes, there is an issue that right now boost doesn't build for ARM due to the bug in CMake version. PR #7146 addresses that, but it's not merged yet (still needs some approvals).
there is no problem with approvaltests, for now the best you can do is to skip ARM builds for a while (until boost PR is merged):
```
if self.settings.arch == "armv8":
    raise ConanInvalidConfiguration("see #7146")
```

**Comment by SSE4** [comment]
okay, #7146 has been merged, so packages should be there. I've re-triggered the CI.

**Comment by madebr** [comment]
Looks like cancelled builds are interpreted by conan-center-bot as successfully.

**Comment by SSE4** [comment]
> Looks like cancelled builds are interpreted by conan-center-bot as successfully.

right, seems to be another bug! thanks for spotting!

**Comment by claremacrae** [comment]
Is there anything we need to do?

**Comment by madebr** [comment]
You need to add `"arch"` to `settings`.

**Comment by claremacrae** [comment]
> You need to add `"arch"` to `settings`.

Thanks... Our release process tests that the PR builds... we didn't think to do that with the earlier change today - but I did then test the most recent commit, so hopefully it will all be good now.

**Comment by SSE4** [comment]
> > You need to add `"arch"` to `settings`.
> 
> Thanks... Our release process tests that the PR builds... we didn't think to do that with the earlier change today - but I did then test the most recent commit, so hopefully it will all be good now.

we recently added some changes which enhanced test coverage for header only packages.
previously, header-only packages were built only on Linux.
now, they at least built once on Windows/macOS/Linux.
it may cause some recipes which previously were green to fail now.

**Comment by madebr** [comment]
This is a fluke.
boost/1.77.0 was just merged; https://github.com/conan-io/conan-center-index/pull/6929

2 things you can do:
bump the boost version used by approvaltests.cpp,
or close and re-open this pr.

**Comment by claremacrae** [comment]
> This is a fluke.
> boost/1.77.0 was just merged; #6929
> 
> 2 things you can do:
> bump the boost version used by approvaltests.cpp,
> or close and re-open this pr.

Thanks @madebr - I cannot close and re-open the PR as it was created by the other co-author of approval tests, and we have finished pairing for today...

So please could someone on the Conan team close and re-open this?

**Comment by claremacrae** [comment]
I'm curious... Why would the merging of boost 1.77.0 cause builds here to fail?

**Comment by madebr** [comment]
https://github.com/conan-io/conan-center-index/pull/6929 caused all boost packages to be rebuilt.
I assume pushing a new recipe + package to conan central is not an atomic action.

**Comment by SSE4** [comment]
> #6929 caused boost packages of all versions to be rebuilt.
> I assume pushing a new recipe + package to conan central is not an atomic action.

yes, seems to be a case. probably, not a transactional operation, e.g. recipe is already updated on remote, but not all packages at the same time are available.

**Review by madebr — APPROVED** [review]

**Review by SSE4 — DISMISSED** [review]

File: recipes/approvaltests.cpp/all/conanfile.py
Comment (prince-chrismc) [inline recipes/approvaltests.cpp/all/conanfile.py:87]:
```suggestion
            raise ConanInvalidConfiguration("Currently unsupported by this recipe, see https://github.com/conan-io/conan-center-index/pull/7213#issuecomment-916983425 for more information")
```

**Review by prince-chrismc — COMMENTED** [review]
Hate to waste two good approvals but future us will not remember the comment is out of date...

I got confused since it was merged and the problem is not releated to the link anymore

File: recipes/approvaltests.cpp/all/conanfile.py
Comment (madebr) [reply recipes/approvaltests.cpp/all/conanfile.py:87]: Perhaps that asm should be replaced with `abort()`

Comment (SSE4) [reply recipes/approvaltests.cpp/all/conanfile.py:87]: they need to merge this fix: https://github.com/catchorg/Catch2/commit/a25c1a24af8bffd35727a888a307ff0280cf9387#diff-ac110b2b11e08efa102d4b37f6a479fb006cdfb9822af039ee07021b349e95bb
btw, it handles too few architectures right now (armv8 and x86/x86_64). maybe something like:
```
#if defined(_MSC_VER)
    DebugBreak();
#elif defined(__GNUC__)
    __builtin_trap();
#else
    abort();
#endif
```
to be more portable.

**Comment by claremacrae** [comment]
> Hate to waste two good approvals but future us will not remember the comment is out of date...
> 
> I got confused since it was merged and the problem is not releated to the link anymore

Thank you very much indeed! I'll follow that PR and try to remember to remove this check when it's no longer needed...

**Comment by claremacrae** [comment]
I just wanted to capture what I've found out...

> it seems to be using some x86_64 ASM even if compiling for ARM?
> 
> ```
> #define CATCH_TRAP() __asm__("int $3\n" : : ) /* NOLINT */
> ```

That code is in Catch2... 

This ApprovalTests.cpp Conan recipe is currently using Catch2 2.11.0. The latest is 2.13.7 and that has lots of improvements to catch_debugger.h, very likely fixing this problem..

**QUESTION - which of these should I do** 

1. Should we leave this current PR as it is, and then I'll update the recipe to use the newer Catch2 in a separate PR 
2. Or should I update this PR to update the Catch2 version and remove the check for arm - so fix it all in this one step?

File: recipes/approvaltests.cpp/all/conanfile.py
Comment (claremacrae) [reply recipes/approvaltests.cpp/all/conanfile.py]: Thank you - this helped me see that the problem is because this ApprovalTests.cpp recipe is depending on an old Catch2 version - I think the latest Catch2 v2.x release should work. I've asked what to do about that in a separate comment.

**Comment by SSE4** [comment]
> **QUESTION - which of these should I do**
> 
> 1. Should we leave this current PR as it is, and then I'll update the recipe to use the newer Catch2 in a separate PR
> 2. Or should I update this PR to update the Catch2 version and remove the check for arm - so fix it all in this one step?
> 3. In a separate PR, I could update all the dependency versions, not just Catch2...

up to you, if you want to wait a bit more. solution 2 is ideal for consumers, as they will get ApprovalTest supporting ARMv8 and other architectures.

**Comment by claremacrae** [comment]
> up to you, if you want to wait a bit more. solution 2 is ideal for consumers, as they will get ApprovalTest supporting ARMv8 and other architectures.

Good point - I've just pushed the change...

**Comment by SSE4** [comment]
seems like `doctest` suffers the same issue as `catch` (are they copy-pasting each other 👀 ?):
```
#define DOCTEST_BREAK_INTO_DEBUGGER() __asm__("int $3\n" : :)
```
need this commit: https://github.com/onqtam/doctest/commit/3c5e0fa8d2cc7e6409de5e4e6df8e52fb9627a26
(versions 2.4.1 - 2.4.6 have it, you're using 2.3.6 right now)

**Comment by claremacrae** [comment]
> seems like `doctest` suffers the same issue as `catch` (are they copy-pasting each other 👀 ?):
> 
> ```
> #define DOCTEST_BREAK_INTO_DEBUGGER() __asm__("int $3\n" : :)
> ```
> 
> need this commit: [onqtam/doctest@3c5e0fa](https://github.com/onqtam/doctest/commit/3c5e0fa8d2cc7e6409de5e4e6df8e52fb9627a26)
> (versions 2.4.1 - 2.4.6 have it, you're using 2.3.6 right now)

HUGE THANKS for this information - much appreciated! And I've done that...

**Comment by SSE4** [comment]
```
/home/conan/w/BuildSingleReference/.conan/data/approvaltests.cpp/8.1.0/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/ApprovalTests.hpp:741:14: note:   no known conversion for argument 1 from ‘const doctest::String’ to ‘const string& {aka const std::__cxx11::basic_string<char>&}’
```
did doctest change their interfaces? 🤔

**Comment by SSE4** [comment]
checking again, it's only old versions of `approvaltests` fails with up-to-date `doctest`.
perhaps, only `approvaltests.cpp/8.2.0+` are compatible with new interfaces.

**Comment by claremacrae** [comment]
> checking again, it's only old versions of `approvaltests` fails with up-to-date `doctest`.
> perhaps, only `approvaltests.cpp/8.2.0+` are compatible with new interfaces.

Thank you...

I am feeling a little bit unlucky with this PR - they aren't usually this hard!!! :-)

**Comment by SSE4** [comment]
> > checking again, it's only old versions of `approvaltests` fails with up-to-date `doctest`.
> > perhaps, only `approvaltests.cpp/8.2.0+` are compatible with new interfaces.
> 
> Thank you...
> 
> I am feeling a little bit unlucky with this PR - they aren't usually this hard!!! :-)

yeah, really a bad timing, we upgraded several things in infrastructure which caused previously green builds to fail.
we're trying to do our best to minimize breakage moving forward (adding more platforms)

**Comment by claremacrae** [comment]
> yeah, really a bad timing, we upgraded several things in infrastructure which caused previously green builds to fail.
> we're trying to do our best to minimize breakage moving forward (adding more platforms)

I would like to ask the conan team to see if a way can be found to de-couple the “contributor submitting a PR for an update” from “the contributor discovering the existing build got broken (or a pre-existing limitation was exposed) because of improvements to infrastructure”…

For example, perhaps have a weekly job that re-builds every recipe on the latest infra - and somehow capture which ones failed and notify maintainers…

That way, we could have had separate PRs for “get this library working on new platforms” and “update this library for a new release…”

By the current coupling together of “only find out your previously working recipe is broken when you try to add a new release”, I feel it becomes much harder to deal with…

File: recipes/approvaltests.cpp/all/conanfile.py
Comment (prince-chrismc) [inline recipes/approvaltests.cpp/all/conanfile.py:52]:
```suggestion
            self.requires("doctest/2.4.6" if tools.Version(self.version) >= "8.2.0" else "doctest/2.3.6")
```

**Review by prince-chrismc — COMMENTED** [review]
Perhaps this is a candidate for remove older versions?

File: recipes/approvaltests.cpp/all/conanfile.py
Comment (claremacrae) [reply recipes/approvaltests.cpp/all/conanfile.py:52]: Thank you very much. I thought about a condition like that, but was assuming that the older versions would still fail the CI for armv8 (or whatever the problem platform was).

Do you mean to drop support for older versions of ApprovalTests.cpp @prince-chrismc - I didn't know it was possible to do that...

If so, it solves all these CI problems in stroke, without making the recipe more complex, so is a great idea...

**Comment by claremacrae** [comment]
Things I have found from experimenting on my Mac:

doctest/2.3.6

* All versions of ApprovalTests.cpp in Conan build with doctest/2.3.6 (which is the version referenced in CCI master)
* But that doctest does not build on archv8

doctest/2.4.3

* ApprovalTests.cpp 8.0.0 to 8.8.0 do not build with this
* 8.8.1 and later do build
* I suspect this may be the first doctest release that works on armv8

doctest/2.4.6

* ApprovalTests.cpp 8.0.0 to 8.8.0 do not build with this
* 8.8.1 and later do build

**Comment by claremacrae** [comment]
So there are two orthogonal things going on here...

* Which versions of doctest are compatible with which of ApprovalTests.cpp
* Which versions of doctest are compatible with armv8 builds

I think the fundamental blockers to this PR are:

* Conan added supported for CI platforms that are not compatible with current versions of ApprovalTests.cpp dependencies
* doctest added a breaking change without bumping the major version number...

One of my options is this - I think it's my simplest option, and retains the previous behaviour on default...

* Revert back to previous behaviour of selecting doctest/2.3.6
* Reinstating the refusal to ever run on armv8

**Comment by claremacrae** [comment]
Another of my options is to remove the following versions from the ApprovalTests.cpp repo...

```
8.0.0
8.1.0
8.1.1
8.2.0
8.3.0
8.4.0
8.5.0
8.6.0
8.7.0
8.7.1
8.8.0
```

The following versions would remain:

```
8.8.1
8.9.0
8.9.1
8.9.2
9.0.0
10.0.0
10.0.1
10.1.0
10.1.1
10.1.2
10.2.0
10.2.1
10.3.0
10.4.0
10.5.0
10.5.1
10.6.0
10.7.0
10.7.1
10.8.0
10.9.1
10.10.0
10.11.0
10.12.0
```

Since ApprovalTests.cpp does try hard to honour semantic versioning, and Conan would still support several 8.x.y releases, this maybe wouldn't inconvenience too many users - but if they have pinned to a particular ApprovalTests.cpp release, I think it would still break their builds???

**Comment by claremacrae** [comment]
> One of my options is this - I think it's my simplest option, and retains the previous behaviour on default...
> 
> * Revert back to previous behaviour of selecting doctest/2.3.6
> * Reinstating the refusal to ever run on armv8

Given that ApprovalTests.cpp is used in Conan as a single-header, this should not even prevent uses from accessing the library via Conan on armv8, I presume...

**Comment by madebr** [comment]
You can conditionally add different versions of requirements + do tests.
Something like the following (untested):
```python
def requirements(self):
    if tools.Version(self.version) > "8.8.0":
        self.requires("doctest/2.4.6")
    else:
        self.requires("doctest/2.3.6")

def validate(self):
    if tools.Version(self.version) < "8.0.0":
        if tools.Version(self.deps_cpp_info["doctest"].version) >= "2.4.0":
            raise ConanInvalidConfiguration("The version of doctest is too new, use an older version (pre 2.4.0)") 
```

Then, you can remove some old versions, while keeping some.
With the cmake recipe, we only keep the highest minversion. (e.g. remove 18.0, 18.1, 18.2 but keep 18.3)

**Comment by claremacrae** [comment]
> ```python
> 2.3.6
> ```

Thank you for the suggestion. The problem with this is that doctest 2.3.6 will fail to build on CCI's armv8 CI builds...

**Comment by prince-chrismc** [comment]
Removing versions does not impact any user AFAIK, they just wont get updates anymore for improvements (or new platform support) assuming they are not using revisions.

I would like to suggest 
- pick the highest doctest version possible (my suggest above)
- block arm for all versions less than 8.8.0 (with a nice comment)

I think that would be the best middle ground of the blockers you very nicely summarized

**Comment by madebr** [comment]
What ⬆️ said.
My suggestion would complicate the recipe and make future maintenance harder.

**Comment by claremacrae** [comment]
> Removing versions does not impact any user AFAIK, they just wont get updates anymore for improvements (or new platform support) assuming they are not using revisions.

"assuming they are not using revisions." ... I don't know the consequences of this, if they were using revisions...

> I would like to suggest
> 
> * pick the highest doctest version possible (my suggest above)
> * block arm for all versions less than 8.8.0 (with a nice comment)
> 
> I think that would be the best middle ground of the blockers you very nicely summarized

Ok, so if I understand this correctly, it will then allow me to retain all the earlier versions...

Thank you very much indeed. I will go ahead and do this, but probably tomorrow...

**Review by uilianries — APPROVED** [review]
Less is more

**Comment by claremacrae** [comment]
From @uilianries on #include Discord, I got a better understanding of the (lack of) consequences of removing old versions from the yaml, so I've kept the latest of each on the 8.x and 9.x series, and the most recent 4 on 10.x series...

As I understand it:

* all the deleted releases will still appear in the web site and by conan search command (I was unaware that this would happen)
* these should all just pass the CI with the current rules on this branch 
* and anyone using the deleted releases will get the previous recipes, which will give them the older doctest version, that works with all the ApprovalTests releases...

This massively simplifies our Python recipe...

Fingers crossed the CI just works now...

**Comment by claremacrae** [comment]
> Fingers crossed the CI just works now...

And whatever happens, I'll be warming the planet just a little bit less, with fewer CI build steps in this and future PRs!!!

**Review by prince-chrismc — APPROVED** [review]
LGTM 🚀

**Review by madebr — APPROVED** [review]

**Comment by claremacrae** [comment]
Thank you everyone for all the support on this PR!

**Pull request merged**
